tests/unit/test_prefixmapping.py

```diff
@@ -18,7 +18,94 @@
 
 from mock import MagicMock as Mock
 
-from oio.directory.meta0 import Meta0PrefixMapping
+from oio.directory.meta0 import \
+        Meta0PrefixMapping, \
+        generate_short_prefixes as prefixes, \
+        _slice_services, _bootstrap
+
+
+class TestMeta0Bootstrap(unittest.TestCase):
+
+    def setUp(self):
+        super(TestMeta0Bootstrap, self).setUp()
+
+    def generate_services(self, count, nb_sites=1, fill_token=2):
+        assert(count > 0)
+        assert(nb_sites > 0)
+        sites = ['s' + str(i) for i in range(nb_sites)]
+
+        def _loc(i):
+            s = list()
+            s.append(sites[i % len(sites)])
+            for _ in range(fill_token):
+                s.append('0')
+            s.append('h' + str(i))
+            return '.'.join(s)
+
+        for i in range(count):
+            yield {'addr': '127.0.0.1:{0}'.format(6000+i),
+                   'id': 'srv-{0}'.format(i),
+                   'type': 'meta1', 'tags': {'tag.loc': _loc(i)}}
+
+    def _test_ok(self, groups, sites=1, replicas=1, level=0, fill_token=0):
+        nb_services = sites * 3
+        print "srv =", nb_services, "sites =", sites, "repli =", replicas
+        srv = list(self.generate_services(nb_services, nb_sites=sites,
+                                          fill_token=fill_token))
+        print repr(srv)
+        _after = _bootstrap(srv, groups, replicas, level)
+        ideal_per_slice = (len(groups) * replicas) / sites
+        for start, end in _slice_services(_after, level):
+            total = sum(len(s['bases']) for s in _after[start:end])
+            # Check the sites are evenly loaded
+            bounds = (ideal_per_slice, ideal_per_slice+1)
+            self.assertIn(total, bounds)
+            # Check services are uniformly balanced within the
+            # current slice.
+            ideal_per_node = total / (end - start)
+            bounds = (ideal_per_node, ideal_per_node+1)
+            for s in _after[start:end]:
+                self.assertIn(len(s['bases']), bounds)
+
+    def test_bootstrap_not_enough_sites(self):
+        for groups in (prefixes(1),  # 16
+                       prefixes(2),  # 256
+                       prefixes(3),  # 4096
+                       prefixes(4)):  # 64ki
+            groups = list(groups)
+            for replicas in range(2, 8):
+                for sites in range(1, replicas):
+                    srv = self.generate_services(sites*3, nb_sites=sites)
+                    self.assertRaises(Exception, _bootstrap, srv, groups,
+                                      replicas=replicas, level=0, fill_token=2)
```

Comment:
Here we should expect a specific exception type, to ensure the API does not change silently.